katdal/chunkstore_s3.py

```diff
@@ -112,16 +112,94 @@ def read_array(fp):
+
+    Raises
+    ------
+    :exc:`InvalidToken`
+        If the token is malformed or truncated, or has expired
+    """
+    # A valid JWS Compact Serialization has three segments
+    try:
+        encoded_header, encoded_payload, encoded_signature = token.split('.')
+    except ValueError:
+        raise InvalidToken(token, "Token does not have exactly two dots ('.') "
+                                  "as expected of JWS (maybe it's truncated?)")
+    # Remove signature to avoid cryptic PyJWT error message ("Invalid crypto padding")
+    token_without_sig = '{}.{}.'.format(encoded_header, encoded_payload)
+    try:
+        header = jwt.get_unverified_header(token_without_sig)
+    except jwt.exceptions.DecodeError as err:
+        raise_from(InvalidToken(token, "Could not decode token - maybe it's truncated "
+                                       "or corrupted? ({})".format(err)), err)
+    # Check signature length for typical MeerKAT signature algorithm
+    if header.get('alg') == 'ES256':
+        len_sig = len(encoded_signature)
+        if len_sig != 86:   # 64 bytes when decoded
+            msg = 'Encoded signature has {} bytes instead of 86 - token string ' \
+                  'is too {}'.format(len_sig, 'short' if len_sig < 86 else 'long')
+            raise InvalidToken(token, msg)
+    # Extract token claims and verify everything except signature and audience
+    options = {'verify_signature': False, 'verify_aud': False}
+    try:
+        claims = jwt.decode(token, options=options)
+    except jwt.exceptions.DecodeError as err:
+        # This covers e.g. bad characters in the signature or non-JSON-dict payload
+        raise_from(InvalidToken(token, "Could not decode token - maybe it's truncated "
+                                       "or corrupted? ({})".format(err)), err)
+    except jwt.exceptions.InvalidTokenError as err:
+        raise_from(InvalidToken(token, str(err)), err)
+    return claims
+
+
 class _BearerAuth(requests.auth.AuthBase):
     """Add bearer token to authorisation request header."""
 
     def __init__(self, token):
-        # Character set from RFC 6750
-        if not re.match('^[A-Za-z0-9-._~+/]*$', token):
-            raise ValueError('Token contains invalid characters')
+        self._claims = decode_jwt(token)
+        if 'prefix' not in self._claims:
+            raise InvalidToken(token, "Token has no 'prefix' claim")
         self._token = token
 
     def __call__(self, r):
+        # Check if token authorises URL even before hitting server for better reporting
+        path = urllib.parse.urlparse(r.url).path.lstrip('/')
+        valid_prefixes = self._claims['prefix']
+        if not any(path.startswith(prefix) for prefix in valid_prefixes):
+            allowed = ', '.join(prefix + '*' for prefix in valid_prefixes)
+            raise InvalidToken(self._token, 'Token does not grant permission for '
+                                            '{}, only for {}'.format(path, allowed))
```

Comment:
In the Jenkins coverage report it seems that this check is never tested.

Reply:
This is now also tested by replacing the allow-all token in the tests with a bucket-specific one.